examples_tests/20.Megatexture/main.cpp

```diff
@@ -931,51 +585,71 @@ int main()
     auto gpumesh = driver->getGPUObjectsFromAssets(&mesh_raw, &mesh_raw+1)->front();
 
     core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> gpu_ds2layout = driver->getGPUObjectsFromAssets(&ds2layout.get(),&ds2layout.get()+1)->front();
-    auto gpu_ds2 = driver->createGPUDescriptorSet(std::move(gpu_ds2layout));
-    core::smart_refctd_ptr<video::IGPUBuffer> gpuPrecomputedStuffSsbo;
+    auto gpuds2 = driver->createGPUDescriptorSet(std::move(gpu_ds2layout));
+    core::smart_refctd_ptr<video::IGPUBuffer> ds2_ssbo;
     {
         constexpr uint32_t ENTRY_SZ = sizeof(uint32_t);
         constexpr uint32_t ENTRY_ALIGNMENT = sizeof(uint32_t);//change to 16 if std140
         constexpr uint32_t PRECOMPUTED_STUFF_COUNT = 3u;//pgtab_sz_log2, phys_pg_tex_sz_rcp, vtex_sz_rcp
-        constexpr uint32_t PRECOMPUTED_STUFF_SZ = ETP_COUNT*ENTRY_ALIGNMENT;
-        constexpr uint32_t UBO_SZ = PRECOMPUTED_STUFF_COUNT*PRECOMPUTED_STUFF_SZ;
-        uint8_t ubodata[UBO_SZ]{};
-        uint32_t* ptr = reinterpret_cast<uint32_t*>(ubodata);
+        constexpr uint32_t PRECOMPUTED_STUFF_SZ = VT_COUNT*ENTRY_ALIGNMENT;
+        constexpr uint32_t PRECOMPUTED_SSBO_SZ = PRECOMPUTED_STUFF_COUNT*PRECOMPUTED_STUFF_SZ;
+        uint8_t precomputedStuff[PRECOMPUTED_SSBO_SZ]{};
+        uint32_t* ptr = reinterpret_cast<uint32_t*>(precomputedStuff);
         uint32_t offset = 0u;
         //pgtab_sz_log2
-        for (uint32_t i = 0u; i < ETP_COUNT; ++i)
-            ptr[i*ENTRY_ALIGNMENT/ENTRY_SZ] = core::findMSB(texPackers[i]->getPageTable()->getCreationParameters().extent.width);
+        for (uint32_t i = 0u; i < VT_COUNT; ++i)
+            ptr[i*ENTRY_ALIGNMENT/ENTRY_SZ] = core::findMSB(gpuvt->getPageTable()->getCreationParameters().extent.width);
         ptr += PRECOMPUTED_STUFF_SZ/ENTRY_SZ;
         //phys_pg_tex_sz_rcp
-        for (uint32_t i = 0u; i < ETP_COUNT; ++i)
+        for (uint32_t i = 0u; i < VT_COUNT; ++i)
         {
-            const double f = 1.0 / static_cast<double>(texPackers[i]->getPhysicalAddressTexture()->getCreationParameters().extent.width);
+            const auto& storageImg = gpuvt->getFloatViews().begin()[i]->getCreationParameters().image;
+            const double f = 1.0 / static_cast<double>(storageImg->getCreationParameters().extent.width);
             reinterpret_cast<float*>(ptr)[i*ENTRY_ALIGNMENT/ENTRY_SZ] = f;
         }
         ptr += PRECOMPUTED_STUFF_SZ/ENTRY_SZ;
         //vtex_sz_rcp
-        for (uint32_t i = 0u; i < ETP_COUNT; ++i)
+        for (uint32_t i = 0u; i < VT_COUNT; ++i)
         {
             double f = 1.0;
-            f /= static_cast<double>(texPackers[i]->getPageTable()->getCreationParameters().extent.width);
-            f /= static_cast<double>(texPackers[i]->getPageSize());
+            f /= static_cast<double>(gpuvt->getPageTable()->getCreationParameters().extent.width);
+            f /= static_cast<double>(gpuvt->getPageExtent());
```

Comment:
the `IVirtualTexture` should have a method to either hand you a const reference to a struct that matches 1:1 with this SSBO data or write the data to a `void*`

Reply:
struct is out of question because `phys_pg_tex_sz_rcp` array size is variable. I'm making a writing function, but i wonder how user could know layout of outputted data?

Reply:
meeeh, actually i think it would be best for IVirtualTexture to provide a template struct<storageCount> for user to fill by himself. Or make UBO with precomputed data come in same DS as pagetable and views (having declaration in builtin descriptors.glsl, etc, etc), this is too scattered

Reply:
> 
> 
> struct is out of question because `phys_pg_tex_sz_rcp` array size is variable. I'm making a writing function, but i wonder how user could know layout of outputted data?

Well I thought of this: 

> 
> 
> You could of course store `phys_pg_tex_sz_rcp[3][MAX_FORMAT_VIEWS_OF_FORMAT_META_CLASS]` but it would be a bit unwieldy because you'd need to do:
> 
> ```glsl
> float irr_glsl_VT_getPhysPgTexSzRcp(in uint formatMetaclass, uint in formatID)
> {
>    return phys_pg_tex_sz_rcp[formatMetaclass][formatID];
> }
> ```
> 
> For an easier solution you could put the array in an SSBO together with the LUT and store `phys_pg_tex_sz_rcp[PAGE_TABLE_LAYERS]`
> 
> ```glsl
> float irr_glsl_VT_getPhysPgTexSzRcp(uint in pageTableLayer)
> {
>    return phys_pg_tex_sz_rcp[pageTableLayer];
> }
> ```

you know the uppper bound on layer amount, its 256, so you can have the `uint` format mapping and the `float` recriproval physical suze

Reply:
then you could return a combined struct of two 256 element arrays, one of uint and one of float as "the LUT"

Reply:
> ```
> float irr_glsl_VT_getPhysPgTexSzRcp(uint in pageTableLayer)
> {
>   return phys_pg_tex_sz_rcp[pageTableLayer];
> }
> ```

not possible to calculate such array because layer->sampler_index mapping doesnt give you info about which sampler array (float/int/uint) index it is. Only user can know that

Reply:
this array is `STexData.vtLayer -> Rcp size mapping` build the array in `IVirtualTexture`